### DISCUSSION TITLE
How to prevent some characters from getting encoded?

### DISCUSSION THREAD

**Discussion by babakfp** [rehypejs/rehype#167] [Q&A] [answered]
Hi

### Code

```ts
import { unified } from "unified"
import remarkParse from "remark-parse"
import remarkRehype from "remark-rehype"
import rehypeStringify from "rehype-stringify"

export const transformer = async (markdown: string) => {
    const processor = unified()

    processor.use(remarkParse)

    processor.use(remarkRehype, {
        allowDangerousHtml: true,
    })

    processor.use(rehypeStringify, {
        allowDangerousCharacters: true,
        allowDangerousHtml: true,
    })

    processor.use(() => {
        return (node) => {
            console.log(JSON.stringify(node, null, 4)) // No characters are not encoded
        }
    })

    const result = await processor.process(markdown)

    console.log(result.value) // Some characters are encoded (and I don't know why!)

    return result.value
}
```

### Input

`````md
`bg-radial-<color>`

```js
{
    /* <plugin-name> */
}
```
`````

### Output

```html
<p><code>bg-radial-&#x3C;color></code></p>
<pre><code class="language-js">{
    /* &#x3C;plugin-name> */
}
</code></pre>
```

**Comment by ChristianMurphy** [comment]
Welcome @babakfp! 👋 
Sorry you ran into some confusion.

Rehype needs to encode the `<` in `<plugin-name>` because you used it in a code block, so it is plaintext.
If it were not encoded, it would turn into a custom element (and an invalid one as it has no closing tag).
That behavior of turning known plaintext into HTML would be highly unexpected and insecure.

**Reply by babakfp** [reply]
Hi, Thanks!

![image](https://github.com/rehypejs/rehype/assets/44144724/ce859c6e-49d5-4ff7-ad4e-6b9a31c67c6a)

Can I disable this behavior and handle it from my side?

Because of my use case, I need to handle every string inside a code block as plain text to prevent errors thrown by another tool. After converting things to HTML, the result goes through Svelte and it will treat some of the content as actual code.

**Reply by wooorm** [reply]
You should not. That's how html works.

You're asking about an XY problem. You think you need to solve the encoding but you don't. There's an actual problem you have but you need to tell us more about it

**Reply by babakfp** [reply]
I tried a different approach, but this time the issue was that "rehype-stringify" was adding its serialization on top of mine, which broke things.
I would like to have more control over things. There is this [library](https://www.npmjs.com/package/@starptech/prettyhtml-hast-to-html) that does what I'm looking for, however, it's not been updated for 5 years.

**Reply by ChristianMurphy** [reply]
@babakfp that doesn't answer the main question, what is the problem you are trying to solve?

Rehype's goal is to produce valid HTML.
The example of your desired result is invalid and therefore a non goal for the project.

That said we still may be able to help you solve your issue.
Are you trying to compile for a non-HTML language? Or something else entirely?

**Reply by babakfp** [reply]
I'm attempting to serialize certain characters in Svelte syntax to prevent Svelte from interpreting the code within the block. However, the methods I've tried for serialization may be incorrect:

One approach involves defining an array of character replacements:

```js
const entries = [
    // [/</g, "&lt;"],
    // [/>/g, "&gt;"],
    [/{/g, "&#123;"],
    [/}/g, "&#125;"],
]
```

Another method involves a function to sanitize strings:

```ts
export const sanitizeString = (string: string) => {
    let result = ""
    for (const char of string) {
        // Convert each character to its Unicode representation
        const unicodeChar = char.charCodeAt(0).toString(16).toUpperCase()
        // Concatenate the Unicode representation to the result
        result += "&#x" + unicodeChar + ";"
        console.log(`"${result}"`)
    }
    return result
}
```

However, neither of these approaches seems to work, as the serialized string undergoes unexpected changes when processed by the rehype stringify plugin. For instance, `&#123;` (representing `{`) transforms into `&#x26;#123;`.

**Reply by ChristianMurphy** [reply]
> Svelte syntax to prevent Svelte from interpreting the code within the block

Two thoughts:
1. You may want to ask Svelte about this? It feels like that is an issue if it is intrepreting plain text when it should not be.
2. If you want tooling specifically designed for Svelte, you may want to checkout https://mdsvex.pngwn.io/

**Reply by ChristianMurphy** [reply]
Another option worth noting, you could try https://github.com/rehypejs/rehype-react as a renderer.
It also supports JSX compatible runtimes like Svelte.

**Comment by babakfp** [comment, marked as answer]
I think I solved it.

`constants.ts`:

```ts
export const STRINGIFY_ENTITIES_DEFAULT_DANGEROUS_CHARACTERS = [
    '"',
    "&",
    "'",
    "<",
    ">",
    "`",
] as const

export const STRINGIFY_ENTITIES_DEFAULT_SVELTE_DANGEROUS_CHARACTERS = [
    "{",
    "}",
] as const
```

`rehype-sanitize-md-code.ts`:

```ts
import { stringifyEntities } from "stringify-entities"
import { visit } from "unist-util-visit"
import type { Plugin } from "unified"

import {
    STRINGIFY_ENTITIES_DEFAULT_DANGEROUS_CHARACTERS,
    STRINGIFY_ENTITIES_DEFAULT_SVELTE_DANGEROUS_CHARACTERS,
} from "./constants.js"

export const rehypeSanitizeMdCode: Plugin = () => {
    return (tree) => {
        visit(tree, "element", (node) => {
            // @ts-ignore
            if (node.tagName === "code") {
                visit(node, "text", (childNode) => {
                    // @ts-ignore
                    childNode.type = "raw"

                    // @ts-ignore
                    childNode.value = stringifyEntities(
                        // @ts-ignore
                        childNode.value,
                        {
                            subset: [
                                ...STRINGIFY_ENTITIES_DEFAULT_DANGEROUS_CHARACTERS,
                                ...STRINGIFY_ENTITIES_DEFAULT_SVELTE_DANGEROUS_CHARACTERS,
                            ],
                        }
                    )
                })
            }
        })
    }
}
```


---

Thanks for helping 🌸.

**Reply by ChristianMurphy** [reply]
I'd be very cautious mucking with escapes.
There is a fair amount of risk of opening an app up to unescaped insecure content and causing unexpected errors from characters going through.

I'd encourage you @babakfp and future readers to consider the alternatives noted in https://github.com/orgs/rehypejs/discussions/167#discussioncomment-8815787 and https://github.com/orgs/rehypejs/discussions/167#discussioncomment-8823022

**Reply by Andersama** [reply]
So to reemphasize this for anyone who runs into this in the future, this is a classic sanitization security problem. The context here is that svelte's syntax expects curly brace characters `{}` for its own use, and the problem is that if other things bring in escaped curly brace characters `&#123;` `&#125;` and `rhype-stringify` sees them, it'll unescape them (because these are considered ***SAFE*** for HTML and rehype deals with HTML). Then suddenly content is being wrapped by `{}`'s as opposed to the escape sequences `&#123;` `&#125;` and depending what is there...ideally svelte just throws an error (there's a fair chance this happens). So what's occurring here is a security violation because `{}`'s are ***DANGEROUS*** characters for svelte. They're are a signal for svelte to take over and do things server side!

So bakafp's approach essentially marks `{}` as dangerous characters. So...as long as his input didn't include `{}`'s for svelte related blocks (this will break svelte blocks), he'll have solved for his expected output. This isn't ideal since this is all to do with parsing. So when and where characters should be escape sequences or not can get very complicated, and essentially there's a mismatch here because rehype wants to handle HTML, but we've been handing it svelte code.

I'd imagine since svelte has its own compiler and parser available there's probably a way to let svelte do the heavy-lifting here for us.

Good news...since `mdsvex` transforms markdown (with html sequences) into svelte, it should be properly handling this kind of problem. So if you've landed here as I have, you probably want to use the npm packages `mdsvex` uses for handling svelte code instead of rehype!

(I'm just pulling these from the readme, these might change later)
- [svelte-parse](https://github.com/pngwn/MDsveX/tree/main/packages/svelte-parse) - Generate a svast AST from a Svelte components.
- [svast](https://github.com/pngwn/MDsveX/tree/main/packages/svast) - An AST specification with accompanying TypeScript definitions.
- [svast-stringify](https://github.com/pngwn/MDsveX/tree/main/packages/svast-stringify) - Turn a svast AST into a Svelte component.
- [svast-utils](https://github.com/pngwn/MDsveX/tree/main/packages/svast-utils) - Utilities for working with a svast tree.
